In [1]:
import pandas as pd
import numpy as np
import altair as alt
import os
# import chromatose as ct

import geopandas as gpd
from geopy.distance import geodesic
from itertools import product
import json

from IPython.display import SVG
import scipy.special

/Users/rf50/opt/anaconda3/lib/python3.9/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


### Read in data and re-define dictionaries

In [2]:
df_model_ij = pd.read_csv('processed_data/DF_MODEL_IJ.csv')
df_model_i = pd.read_csv('processed_data/DF_MODEL_I.csv')
df_model_j = pd.read_csv('processed_data/DF_MODEL_J.csv')

counties = [
    'Los Angeles', 'Orange', 'Riverside', 
    'San Bernardino', 'San Diego', 'Ventura']

d_index_county = {i: county for i, county in enumerate(counties)}
d_county_index = {v:k for k, v in d_index_county.items()}

d_county_fips = {
    'Los Angeles':'037',
    'Orange':'059',
    'Riverside':'065',
    'San Bernardino':'071',
    'San Diego':'073',
    'Ventura':'111',
}
d_fips_county = {v: k for k, v in d_county_fips.items()}

### Plan

We will build $d_{ij}$, $F_i$, $Z_j$ from wages.   
We will solve for baseline equilibrium parameters.   
Then increase $F_{LA}$, and re-solve for equilibrium.   
Compare population, wages, housing, and commuting.

At spatial equilibrium: 
- labor markets clear
- housing markets clear
- commuting flows reflect choices
  households are indifferent across residence-workplace pairs


### Construct $d_{ij}$

In [3]:
# First read in centroids from cleaned data
df_centroid = pd.read_csv('processed_data/DF_CENTROID.csv')
df_centroid['fips'] = df_centroid['fips'].astype(str).str.zfill(3)

pairs = pd.DataFrame(
    list(product(df_centroid.fips, df_centroid.fips)
), columns=["fips_origin", "fips_dest"])

# Merge coordinates for county A
pairs = pairs.merge(df_centroid[['fips','centroid_x','centroid_y']], left_on='fips_origin', right_on='fips')
pairs.rename(columns={'centroid_x':'x_origin','centroid_y':'y_origin'}, inplace=True)
pairs.drop(columns='fips', inplace=True)

# Merge coordinates for county B
pairs = pairs.merge(df_centroid[['fips','centroid_x','centroid_y']], left_on='fips_dest', right_on='fips')
pairs.rename(columns={'centroid_x':'x_dest','centroid_y':'y_dest'}, inplace=True)
pairs.drop(columns='fips', inplace=True)

# Euclidean distance in meters
pairs['distance_m'] = np.sqrt((pairs['x_dest'] - pairs['x_origin'])**2 + (pairs['y_dest'] - pairs['y_origin'])**2)
pairs['distance_mi'] = pairs['distance_m'] / 1609.34
pairs['d_ij'] = pairs['distance_mi']

pairs['county_i_name'] = pairs['fips_origin'].map(d_fips_county)
pairs['county_j_name'] = pairs['fips_dest'].map(d_fips_county)

# pairs stores pair-wise distances between county centroids in miles
pairs = pairs[['county_i_name', 'county_j_name', 'd_ij']]
# pairs.head()

In [4]:
# Using web-widget here, we can estimate the diagonals (within county commute LA-LA commute)
# https://www.census.gov/acs/www/about/why-we-ask-each-question/commuting/
# We can then multiply by some scalar factor
# LA: 30 min commute
# OC: 27 min commute
# San Bernardino: 33 minute commute
# Ventura: 26 min commute
# Riverside: 34 min commute
# San Diego: 26 min commute

# in minutes (median from ACS census widget)
factor = 0.25
d_inter_county_commute = {
    'Los Angeles': 30,
    'Orange': 27,
    'San Bernardino': 33,
    'San Diego': 26,
    'Riverside': 34,
    'Ventua': 26
}

# min per mile (factors suggested by chatgpt)
d_inter_county_commute_factor = {
    'Los Angeles': 2.3,
    'Orange': 2.0,
    'San Bernardino': 1.6,
    'San Diego': 2.0,
    'Riverside': 1.6,
    'Ventua': 1.7
}

for k, v in d_inter_county_commute.items():
    factor = d_inter_county_commute_factor[k]
    pairs.loc[
        (pairs['county_i_name']==k) & 
        (pairs['county_j_name']==k)
    , 'd_ij'] = v / factor

In [5]:
# Merge commuting distance matrix d_ij into dataframe
df_model_ij = pd.merge(
    df_model_ij, pairs, how='left', on=['county_i_name','county_j_name'])

df_model_ij.head()

county_i  county_j   county_i_name county_j_name  L_ij_data        d_ij
0         0         0     Los Angeles   Los Angeles    4429523   13.043478
1         1         0          Orange   Los Angeles     180250   32.571940
2         2         0       Riverside   Los Angeles      53172   74.016584
3         3         0  San Bernardino   Los Angeles     132992   54.615943
4         4         0       San Diego   Los Angeles       6075  105.441086

### Add.l Parameters: $F_i$

F: In estimating wildfire exposure of LA, we think about the approximate proportion of structures that were destroyted by the wildfires. If we make the assumption that 10,000 of 1 million structures were destroyed by the fires:
\begin{align}
&F_i = 0.01, \hspace{0.3em} i = \mathrm{Los Angeles}\\
&F_i = 0.00 \hspace{1em} \mathrm{otherwise} \\
\end{align}

In [6]:
d_Fi_nofire = {
    'Los Angeles':0, 
    'Orange':0, 
    'Riverside':0, 
    'San Bernardino':0,
    'San Diego':0, 
    'Ventura':0
}
d_Fi_fire1 = {
    'Los Angeles':0.01, 
    'Orange':0, 
    'Riverside':0, 
    'San Bernardino':0,
    'San Diego':0, 
    'Ventura':0
}
df_model_i['F_i_nofire'] = df_model_i['county_i_name'].map(d_Fi_nofire)
df_model_i['F_i_fire1'] = df_model_i['county_i_name'].map(d_Fi_fire1)

In [7]:
kappa = 0.08 # Redding 2016: 0.05–0.15 per mile, 3-10% wage loss per 10 miles
chi = 0.5 # Calibrate this: low chi: small fire response, high fire chi: large response
eta = 0.7 # Housing supply elasticity saiz 2010

We can come back to estimate $\kappa$ from the gravity eqns, but for now we will set it to 0.08 and move on
\begin{align}
L_{ij}^{data} = \pi_{ij} N
\end{align}

We have now built $d_{ij}$, $F_i$, $Z_j$ from wages.   
We will solve for baseline equilibrium.   
Then increase $F_{LA}$, and re-solve for equilibrium.   
Compare population, wages, housing, and commuting.

### Back out baseline housing prices

At spatial equilibrium:  
- labor markets clear
- housing markets clear
- commuting flows reflect choices
  households are indifferent across residence-workplace pairs

Housing market clearing:
\begin{align}
p_i = \left(\frac{N_i}{H_i}\right)^{1/\eta}
\end{align}

In [8]:
def compute_prices(N_i, H_i, eta):
    p_i = (N_i / H_i) ** (1/eta)
    return p_i
    
df_model_i['p_i_eq'] = compute_prices(
    df_model_i['N_i_2024_data'], 
    df_model_i['H_i_2024_data'], 
    eta
)

In [9]:
# df_model_i[['county_i_name', 'p_i_eq']].sort_values(by='p_i_eq', ascending=False)

We notice that $p_i$ is not super aligned with median county home prices rankings.
We note that we are unsure whether reported population that is used to estimate number of households include unhoused individuals/families or not.

### Normalize wages w_j for Z_j

In [10]:
def compute_wages_norm(L_j, w_j_emp):
    weights = L_j / L_j.sum()
    w_bar = np.sum(weights * w_j_emp) # weighted avg
    w_j_emp_norm = w_j_emp / w_bar
    return w_j_emp_norm

# normalize weighted avg wage per employee
weights = df_model_j['L_j_data'] / df_model_j['L_j_data'].sum()
w_bar = np.sum(weights * df_model_j['w_j_emp_data']) # weighted avg
df_model_j['w_j_emp_data'] / w_bar

df_model_j['w_j_emp_norm_data'] = compute_wages_norm(
    df_model_j['L_j_data'], df_model_j['w_j_emp_data']
)

w_per_capita = df_model_j['w_j_county_data'] / df_model_j['L_j_data']
w_bar_capita = np.sum(weights * w_per_capita)
df_model_j['w_j_county_norm_data'] = w_per_capita / w_bar_capita

### Back out Amenities
By definition we have a relation between $L_i$, working residents living in county i, and $\pi_{ij}$, the choice probabilities of residents living in county i and working in county j. We can treat both the denominator sum on the r.h.s. and N on the l.h.s. as a fixed constant C, and solve for amenities $A_i$ up to that fixed constant.

\begin{align}
L_i &= L \displaystyle \sum_{j} \pi_{ij}\\
\cfrac{L_i}{L} &= \displaystyle \sum_{j} \cfrac{\exp ({\log w_j - \kappa d_{ij} - \log p_i + A_i - \chi F_i) }}{ \sum_k \exp ({\log w_k - \kappa d_{ik} - \log p_i + A_i - \chi F_i) }}\\
A_i &= \log {L_i}  - \log \left( \sum_j \exp(\log w_j - \kappa d_{ij} - \log P_i - \chi F_i) \right) + C
\end{align}

When we are solving for baseline, the $F_i$ vector is 1.

Something to be careful of: have to be consistent with $N_{i}$ and $L_{i} = \sum_j L_{ij}$. We do have both datasets:

In [11]:
df_model_i[['county_i', 'N_i_2024_data', 'L_i_data']].sort_values(
    by='N_i_2024_data', ascending=False) / 1e6

_ = df_model_i.copy()
_['N_i_2024_mils'], _['L_i_mils'] = _['N_i_2024_data']/1e6, _['L_i_data']/1e6
_['workers_per_household'] = _['L_i_data']/_['N_i_2024_data']
_[['county_i_name', 'N_i_2024_mils', 'L_i_mils', 'workers_per_household']]

county_i_name  N_i_2024_mils  L_i_mils  workers_per_household
0     Los Angeles       3.304914  4.742279               1.434918
1          Orange       1.056690  1.547549               1.464525
2       Riverside       0.835918  1.016166               1.215628
3  San Bernardino       0.698524  0.909384               1.301865
4       San Diego       1.214418  1.615443               1.330220
5         Ventura       0.272429  0.389548               1.429909

We observe that $L_i$ is slightly larger than $N_i$ meaning that slightly more than one person works in each household.

In [12]:
# merge ij, i, j dataframes for \pi_ij computation
df_model_merge = df_model_ij.copy()
df_model_merge = df_model_merge.merge(
    df_model_i, on=['county_i_name','county_i']
    ).merge(df_model_j,on=['county_j_name','county_j'])


# constructing first term (vector in i)
_df_i = df_model_i.copy()
_df_i['_term1'] = np.log(
    df_model_ij.groupby('county_i')['L_ij_data'].sum()
)


# constructing second term 
_df_ij = df_model_merge.copy()
_df_ij['_exp'] = np.exp(
    np.log(df_model_merge['w_j_emp_data']) - \
    kappa * df_model_merge['d_ij'] - \
    np.log(df_model_merge['p_i_eq']) - \
    chi * df_model_merge['F_i_nofire']
)
_df_i['_term2'] = np.log(_df_ij.groupby('county_i')['_exp'].sum())
_df_i['A_i_eq'] = _df_i['_term1'] - _df_i['_term2']
_df_i['A_i_norm_eq'] = _df_i['A_i_eq'] - _df_i['A_i_eq'].mean() # normalize 

if 'A_i_eq' not in df_model_i.columns:
    df_model_i = df_model_i.merge(
        _df_i[['county_i', 'A_i_eq', 'A_i_norm_eq']], on='county_i'
    )

In [13]:
# df_model_i[['county_i_name', 'A_i_norm_eq']
#     ].sort_values(by='A_i_norm_eq', ascending=False)

In [14]:
df_model_merge = df_model_merge.merge(
    df_model_i[['county_i_name', 'A_i_norm_eq']], on='county_i_name')

### Compute baseline $\pi_{ij}$

- We could either do joint probability $\pi_{i, j}$
- or do conditional probability $\pi_{j|i}$
- We model residence–workplace choice as a multinomial logit over all (i,j) pairs. This formulation is equivalent to a two-step model where households choose residence i and then workplace j conditional on residence, as in Rossi-Hansberg (2019).


In [15]:
def stable_softmax(x):
    x_max = np.max(x)
    e_x = np.exp(x - x_max)
    return e_x / e_x.sum()

def compute_U_ij(w_j, p_i, d_ij, A_i, F_i, kappa, chi):
    U_ij = np.log(w_j)  - np.log(p_i) \
        - kappa * d_ij + A_i - chi * F_i
    return U_ij
    
df_model_merge['U_ij_eq'] = compute_U_ij(
    df_model_merge['w_j_emp_norm_data'],
    df_model_merge['p_i_eq'],
    df_model_merge['d_ij'],
    df_model_merge['A_i_norm_eq'],
    df_model_merge['F_i_nofire'],
    kappa, chi
)


# ---- joint probability over all (i,j) ----
df_model_merge['pi_ij_joint_eq'] = stable_softmax(df_model_merge['U_ij_eq'])
# scipy.special.softmax(df_model_merge['U_ij_eq'])

# ---- conditional probability, given residence i ----
# softmax over j within each county_i
df_model_merge['pi_ij_cond_eq'] = \
    df_model_merge.groupby('county_i')['U_ij_eq'].transform(
        # lambda x: scipy.special.softmax(x.values)
        lambda x: stable_softmax(x.values)
        
)
# merge pi's and U back to
df_model_ij = df_model_ij.merge(df_model_merge[
    ['county_i', 'county_j', 'U_ij_eq','pi_ij_cond_eq', 'pi_ij_joint_eq']
], on=['county_i', 'county_j'])

In [16]:
# Assess predicted Lij and Lij data
_ = df_model_merge.copy()

# _['predicted_L_ij_eq'] = _['pi_ij_cond_eq'] * _['N_i_2024_data']
_['predicted_L_ij_eq'] = _['pi_ij_cond_eq'] * _['L_i_data']

_['%_diff_pi_L'] = (_['predicted_L_ij_eq'] - _['L_ij_data']) / _['L_ij_data'] * 100
_[['county_i_name', 'county_j_name', 'L_ij_data', 'predicted_L_ij_eq', 'pi_ij_joint_eq', '%_diff_pi_L']]

county_i_name   county_j_name  L_ij_data  predicted_L_ij_eq  \
0      Los Angeles     Los Angeles    4429523       3.656523e+06   
1           Orange     Los Angeles     180250       2.574428e+05   
2        Riverside     Los Angeles      53172       1.282617e+04   
3   San Bernardino     Los Angeles     132992       4.887121e+04   
4        San Diego     Los Angeles       6075       9.787651e+02   
5          Ventura     Los Angeles      66180       1.116336e+04   
6      Los Angeles          Orange     196169       7.468841e+05   
7           Orange          Orange    1325458       1.153403e+06   
8        Riverside          Orange      75863       9.089658e+04   
9   San Bernardino          Orange      37327       1.031400e+05   
10       San Diego          Orange      12120       1.269541e+04   
11         Ventura          Orange        576       9.971315e+02   
12     Los Angeles       Riverside      16786       1.990215e+04   
13          Orange       Riverside      16208       4.861586e+04   
14       Riverside       Riverside     725518       6.246696e+05   
15  San Bernardino       Riverside      78508       2.003469e+05   
16       San Diego       Riverside       6989       1.924493e+04   
17         Ventura       Riverside        177       2.140608e+01   
18     Los Angeles  San Bernardino      60779       9.810419e+04   
19          Orange  San Bernardino      12934       7.136564e+04   
20       Riverside  San Bernardino     113716       2.591877e+05   
21  San Bernardino  San Bernardino     657249       5.533679e+05   
22       San Diego  San Bernardino       1209       2.126404e+03   
23         Ventura  San Bernardino        586       1.270579e+02   
24     Los Angeles       San Diego       4724       2.234564e+03   
25          Orange       San Diego      12094       9.990533e+03   
26       Riverside       San Diego      47073       2.831573e+04   
27  San Bernardino       San Diego       2796       2.418387e+03   
28       San Diego       San Diego    1588755       1.580363e+06   
29         Ventura       San Diego        333       3.968193e+00   
30     Los Angeles         Ventura      34298       2.186309e+05   
31          Orange         Ventura        605       6.731265e+03   
32       Riverside         Ventura        824       2.701786e+02   
33  San Bernardino         Ventura        512       1.239606e+03   
34       San Diego         Ventura        295       3.404043e+01   
35         Ventura         Ventura     321696       3.772351e+05   

    pi_ij_joint_eq  %_diff_pi_L  
0     3.577682e-01   -17.451088  
1     2.518919e-02    42.825424  
2     1.254962e-03   -75.877953  
3     4.781746e-03   -63.252522  
4     9.576612e-05   -83.888640  
5     1.092266e-03   -83.131822  
6     7.307800e-02   280.735038  
7     1.128533e-01   -12.980806  
8     8.893669e-03    19.816745  
9     1.009161e-02   176.314730  
10    1.242167e-03     4.747592  
11    9.756316e-05    73.113111  
12    1.947302e-03    18.563982  
13    4.756761e-03   199.949764  
14    6.112007e-02   -13.900186  
15    1.960271e-02   155.193034  
16    1.882997e-03   175.360239  
17    2.094453e-06   -87.906170  
18    9.598889e-03    61.411320  
19    6.982688e-03   451.767768  
20    2.535991e-02   127.925436  
21    5.414363e-02   -15.805447  
22    2.080555e-04    75.881252  
23    1.243183e-05   -78.317773  
24    2.186383e-04   -52.697622  
25    9.775119e-04   -17.392653  
26    2.770520e-03   -39.847187  
27    2.366242e-04   -13.505481  
28    1.546288e-01    -0.528184  
29    3.882632e-07   -98.808350  
30    2.139169e-02   537.445107  
31    6.586127e-04  1012.605762  
32    2.643531e-05   -67.211336  
33    1.212878e-04   142.110551  
34    3.330646e-06   -88.460870  
35    3.691012e-02    17.264460

In [17]:
# `%_diff_pi_L` is the percent difference between
# our model’s predicted flow and the observed flow relative to the observed flow

_['%_diff_pi_L'].describe()

count      36.000000
mean       73.946312
std       216.856995
min       -98.808350
25%       -55.336347
50%         2.109704
75%       131.471715
max      1012.605762
Name: %_diff_pi_L, dtype: float64

For most commuting flows, our observed $L_{ij}^{\mathrm{model}} ~= L_{ij}^{\mathrm{data}}$. 50% of our predicted commuting flows are within -55% to 131% off. Let's just move on, and keep in mind that we should always compare counterfactuals to the model's baseline predictions for $L_{\mathrm{ij, predicted baseline}}$ and $L_{\mathrm{ij, counterfactual}}$, not$L_{\mathrm{ij, data}}$ and $L_{\mathrm{ij, counterfactual}}$
Same for $N_i$, 

Once we are in structural / spatial equilibrium mode, many objects become model-implied, not data objects anymore.

In [18]:
# (misleading column names), we are now summing over all j 
# to see if we satisfied our condition that predicted Li = Li data
_.groupby('county_i')[['predicted_L_ij_eq','L_ij_data']].sum()/1e6

predicted_L_ij_eq  L_ij_data
county_i                              
0                  4.742279   4.742279
1                  1.547549   1.547549
2                  1.016166   1.016166
3                  0.909384   0.909384
4                  1.615443   1.615443
5                  0.389548   0.389548

In [19]:
scatter = alt.Chart(_).mark_circle(color='#CD5C5C').encode(
    x=alt.X('L_ij_data', 
            scale=alt.Scale(type='symlog'), 
            title='Observed L_ij'),
    y=alt.Y('predicted_L_ij_eq', 
            scale=alt.Scale(type='symlog'),
            title='Predicted L_ij'),
    tooltip=['county_i_name', 'county_j_name', 'L_ij_data', 'predicted_L_ij_eq']
).properties(
    width=400,
    height=350,
    title='Observed vs Predicted Commuting Flows'
)
# y = x line
max_val = max(_['L_ij_data'].max(), _['predicted_L_ij_eq'].max())
line = alt.Chart(pd.DataFrame({'x':[0, max_val], 'y':[0, max_val]})).mark_line(
    color='#898989', strokeDash=[5,5], strokeWidth=.5
).encode(
    x='x',
    y='y'
)
chart_obs_pred = line + scatter
# chart_obs_pred.save('output/chart_obs_pred_lin.svg')
chart_obs_pred

alt.LayerChart(...)

This chart is mainly a sanity check that our baseline parameters from model inversion are roughly estimating $\pi_{ij} \hspace{0.3em} \mathrm{s.t.} \hspace{0.3em} L_{ij}^{\mathrm{model}} = L_{ij}^{\mathrm{data}}$

### Running the counterfactual

- Baseline equilibrium: ${N_i^{(0)}, L_{ij}^{(0)}, w_j^{(0)}, p_i^{(0)}, \pi_{ij}^{(0)}, }$
- Change $Fi$
- Recompute utilities $U_{ij}^{(1)} =  \log {w_j^{(1)}} - \log {p_i^{(1)}} \kappa d_{ij} + A_i - \chi F_i$ 
- Update choices $\pi_{ij}^{(1)} = \mathrm{softmax_j} (U_{ij}^{(1)})$
- Update implied worker locations $L_i^{(1)} =  L \sum_j \pi_{ij}^{(1)}$
- Update implied populations $N_i^{(1)} = L_i^{(1)}/\lambda_i$
- Update housing prices $p_i^{(1)} = \left( \cfrac{N_i^{(1)}}{H_i^{(1)}} \right)^ \eta$
- Labor supply changes wages: $w_j^{(1)} = Z_j (L_j^{(1)})^ {-\alpha}$

Iterate ^ until $N_i^{(t+1)} - N_i^{(t)}$ converges

In [20]:
def compute_prices(N_i, H_i, eta):
    p_i = (N_i / H_i) ** (1/eta)
    return p_i

def compute_wages_norm(L_j, w_j_emp):
    weights = L_j / L_j.sum()
    w_bar = np.sum(weights * w_j_emp) # weighted avg
    w_j_emp_norm = w_j_emp / w_bar
    return w_j_emp_norm


In [21]:
df_i_cf = df_model_i.copy()          # county-level
df_j_cf = df_model_j.copy()          # county-level
df_ij_cf = df_model_merge.copy()     # commuting matrix

df_i_cf = df_i_cf.rename(columns={
    'N_i_2024_data':'N_i', # initialize values (but they will change in loop)
    'L_i_data':'L_i',      # initialize values (but they will change in loop)
    'p_i_eq':'p_i',        # initialize values (but they will change in loop)
    'A_i_norm_eq':'A_i',
    'H_i_2024_data':'H_i',
})[['county_i_name', 'county_i', 'H_i', 'N_i', 'L_i', 'F_i_fire1', 'p_i', 'A_i']]

df_j_cf = df_j_cf.rename(columns={
    'w_j_emp_data':'w_j_emp', 
    'L_j_data':'L_j',
    'w_j_emp_norm_data':'w_j_emp_norm',
})[['county_j_name', 'county_j', 'w_j_emp','w_j_emp_norm', 'L_j']]

df_ij_cf = df_ij_cf.rename(columns={
    'L_ij_data':'L_ij',
    'U_ij_eq':'U_ij',
    'pi_ij_joint_eq':'pi_ij_joint',
    'pi_ij_cond_eq':'pi_ij_cond',
})[['county_i', 'county_j', 'county_i_name', 'county_j_name', 
    'L_ij', 'd_ij', 'U_ij', 'pi_ij_joint', 'pi_ij_cond']]

# re-merge all 'i' vars (in particular housing has changed)
# this way any counterfactual changes can just be df_i and df_ij will re-merge i-level data
df_ij_cf = df_ij_cf.merge(df_i_cf, how='left', on=['county_i', 'county_i_name'])
df_ij_cf = df_ij_cf.merge(df_j_cf, how='left', on=['county_j', 'county_j_name'])

In [22]:
def solve_equilibrium(
    df_i, df_j, df_ij,
    kappa, 
    chi, # fire coefficient
    structure_shock=0.01,
    max_iter=500, 
    tol=1e-6
):
    df_i = df_i.copy()
    df_j = df_j.copy()
    df_ij = df_ij.copy()

    # Setting housing stock to 0.99 of original
    H_i_eq_LA = df_i.loc[df_i['county_i_name']=='Los Angeles', 'H_i'].values[0]
    df_i.loc[df_i['county_i_name']=='Los Angeles', 'H_i'] = H_i_eq_LA * (1-structure_shock)


    for it in range(max_iter):
        # ---- compute utilities ----
        df_ij['U_ij'] = compute_U_ij(
            df_ij['w_j_emp_norm'],
            df_ij['p_i'],
            df_ij['d_ij'],
            df_ij['A_i'],
            df_ij['F_i_fire1'],
            kappa, chi
        )

        # ---- conditional choice ----
        df_ij['pi_ij_cond'] = df_ij.groupby('county_i')['U_ij'].transform(
            lambda x: stable_softmax(x.values)
        )

        # ---- implied commuting flows ---- # <-- Li or Ni??
        df_ij['L_ij'] = df_ij['pi_ij_cond'] * df_ij['N_i']

        # ---- update employment by workplace ----
        L_j = df_ij.groupby('county_j')['L_ij'].sum().rename('L_j')

        # ---- update prices (housing market clearing) ----
        df_j = df_j.drop(columns=['L_j']) # drop previous values
        df_j = df_j.merge(L_j, left_on='county_j', right_index=True, how='left')
        p_i_new = (df_i['N_i'] / df_i['H_i']) ** eta

        # ---- convergence check ----
        err = np.max(np.abs(p_i_new - df_i['p_i']))
        df_i['p_i'] = 0.7 * df_i['p_i'] + 0.3 * p_i_new  # damping

        if err < tol:
            print(f"Converged in {it} iterations")
            break

    return df_i, df_j, df_ij

The above answers the question: Given a housing shock in LA, how do commuting patterns and prices adjust holding population and wages fixed? This is a short-run equilibrium in which population is fixed and commuting patterns and housing prices adjust endogenously. 



In [23]:
df_i_cf, df_j_cf, df_ij_cf = solve_equilibrium(df_i_cf, df_j_cf, df_ij_cf, kappa, chi=2)

Converged in 32 iterations


/var/folders/6y/m69ngqb568nbb5t5p3lz216r0000gn/T/ipykernel_19879/1389104465.py:15: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '3659440.95' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df_i.loc[df_i['county_i_name']=='Los Angeles', 'H_i'] = H_i_eq_LA * (1-structure_shock)


In [24]:
# only p_i changes in this df
df_compare_i = df_model_i.merge(
    df_i_cf[['county_i_name','county_i', 'N_i', 'L_i', 'p_i']].rename(
        columns={'p_i':'p_i_cf'})[['county_i_name', 'p_i_cf']], how='left', on='county_i_name'
)
df_compare_i['diff_p_i'] = df_compare_i['p_i_cf'] - df_compare_i['p_i_eq']

# only L_j changes in this dataframe
df_compare_j = df_model_j.merge(
    df_j_cf.rename(columns={'L_j':'L_j_cf'}
    )[['county_j_name', 'L_j_cf']], how='left', on='county_j_name'
)
df_compare_j['diff_L_j'] = df_compare_j['L_j_cf'] - df_compare_j['L_j_data']

In [26]:
# Uij changes only for LA-X
# L_ij changes significantly
# pi_ij change very minutely - chat says microscopic chagnes in pi_ij can have huge changes in L_ij
df_compare_ij =df_model_ij.merge(df_ij_cf[['county_i', 'county_j', 'L_ij', 'U_ij','pi_ij_joint', 'pi_ij_cond']].rename(columns={
    'L_ij':'L_ij_cf',
    'U_ij':'U_ij_cf',
    'pi_ij_joint':'pi_ij_joint_cf',
    'pi_ij_cond':'pi_ij_cond_cf'
}), how='left', on=['county_i', 'county_j'])

df_compare_ij = df_compare_ij[[
    'county_i_name', 'county_j_name', 'L_ij_data','L_ij_cf', 
   'U_ij_eq', 'U_ij_cf', 'pi_ij_cond_eq', 'pi_ij_cond_cf', 
   'pi_ij_joint_eq', 'pi_ij_joint_cf'
]]

df_compare_ij['diff_U_ij'] = df_compare_ij['U_ij_cf'] - df_compare_ij['U_ij_eq']
df_compare_ij['diff_pi_ij_joint'] = df_compare_ij['pi_ij_joint_cf'] - df_compare_ij['pi_ij_joint_eq']
df_compare_ij['diff_pi_ij_cond'] = df_compare_ij['pi_ij_cond_cf'] - df_compare_ij['pi_ij_cond_eq']
df_compare_ij['diff_L_ij'] = df_compare_ij['L_ij_cf'] - df_compare_ij['L_ij_data']

# df_compare_ij[['county_i_name','county_j_name', *[x for x in df_compare_ij.columns if 'diff' in x]]]

In [27]:
df_compare_ij.to_csv('processed_data/DF_COMPARE_IJ.csv', index=False)
df_compare_i.to_csv('processed_data/DF_COMPARE_I.csv', index=False)
df_compare_j.to_csv('processed_data/DF_COMPARE_J.csv', index=False)

In [ ]:
# plot changes in p_i, plot L_j, and commuting L_ij differences